src/RawPOSIX/src/safeposix/dispatcher.rs

```diff
@@ -203,79 +203,197 @@ pub fn lind_syscall_api(
+        
+            // Validate and convert user buffer address to system address
+            // NaCl: Uses NaClUserToSysAddrRangeProt with similar protection flags
+            // PROT_READ is correct because write() reads FROM the buffer
+            let buf = match check_and_convert_addr_ext(&cage, arg2, count, PROT_READ) {
                 Ok(addr) => addr as *const u8,
-                Err(errno) => return syscall_error(errno, "write", "invalid buffer address"),
+                Err(errno) => {
+                    return syscall_error(
+                        errno,
+                        "write",
+                        "buffer access violation or invalid address"
+                    );
+                }
             };
+        
+            // Perform write operation through cage abstraction
             cage.write_syscall(fd, buf, count)
         }
 
         WRITEV_SYSCALL => {
+            // NaCl equivalent: NaClSysWritev
             let fd = arg1 as i32;
+            let iovcnt = arg3 as i32;
+        
+            // NaCl validates count first
+            if iovcnt <= 0 {
+                return syscall_error(
+                    Errno::EINVAL,
+                    "writev",
+                    "invalid iovec count"
+                );
+            }
+        
             let cage = interface::cagetable_getref(cageid);
-            let iovec = match check_and_convert_addr_ext(&cage, arg2, size_of::<interface::IovecStruct>(), PROT_READ) {
+        
+            // Validate the iovec array address first
+            // This matches NaCl's validation order
+            let iov_base = match check_and_convert_addr_ext(
+                &cage,
+                arg2,
+                (iovcnt as usize) * std::mem::size_of::<interface::IovecStruct>(),
+                PROT_READ
+            ) {
                 Ok(addr) => addr as *const interface::IovecStruct,
-                Err(errno) => return syscall_error(errno, "writev", "invalid iovec address"),
+                Err(errno) => {
+                    return syscall_error(
+                        errno,
+                        "writev",
+                        "invalid iovec array address"
+                    );
+                }
             };
-            let iovcnt = arg3 as i32;
-            cage.writev_syscall(fd, iovec, iovcnt)
+        
+            // NaCl validates each iovec entry's buffer - we do this in writev_syscall
+            // The actual write operation is delegated to the cage implementation
+            cage.writev_syscall(fd, iov_base, iovcnt)
         }
 
         MUNMAP_SYSCALL => {
+            // NaCl equivalent: NaClSysMunmap
             let addr = arg1 as *mut u8;
-            let len = arg2 as usize;
+            let length = arg2 as usize;
+            let cage = interface::cagetable_getref(cageid);
+        
+            // TODO(Security): Need to implement the following NaCl-style security checks:
```

Comment:
These are already handled in `munmap_handler`

Reply:
Okay, removed